### PR TITLE
Add Lua functions to add and modify spell cooldowns.

### PR DESCRIPTION
**Pull request by Terrorblade**
Add two lua functions for adding and modifying spell cooldowns.

### CODE REVIEW

**Comment by 55Honey** [comment]
All timers seem to be in seconds instead of milliseconds.

```lua
local SPELL_HEARTHSTONE = 8690
local ITEM_HEARTHSTONE = 6948

local function command( event, player, command, chathandler )
    if player then
        if command == 'cddown' then
            player:SendUnitSay('1', 0)
            player:ModifySpellCooldown(SPELL_HEARTHSTONE,-600)
        elseif command == 'cdup' then
            player:ModifySpellCooldown(SPELL_HEARTHSTONE,600)
            player:SendUnitSay('2', 0)
        elseif command == 'addcd' then
            player:AddSpellCooldown(SPELL_HEARTHSTONE, 1800)
            player:SendUnitSay('3', 0)
        end
    end
end

RegisterPlayerEvent(42, command)
```
I tried this. `addcd`  seems to have no effect, but the character is stuck in a cast animation without a castbar happening, if i click the hearhstone after using it.
Using  `cddown` 3 times removes the bug, so something is happening core side.

```lua
player:AddSpellCooldown(SPELL_HEARTHSTONE, 1800, true, ITEM_HEARTHSTONE)
```
did not change anything.

`cdup` and `cddown` work fine, if there's a cooldown currently. If `cdup` exceeds the default cooldown, the hearthstone icon's shade bugs out.

File: src/LuaEngine/PlayerMethods.h [hunk 1]
```diff
@@ -4199,5 +4199,59 @@ namespace LuaPlayer
     player->RemovePet(player->GetPet(), (PetSaveMode)mode, returnreagent);
     return 0;
     }*/
+
+    /**
+     * Starts a spell cooldown for the [Player]
+     *
+     * @param uint32 SpellId : Spell Id to add cooldown to.
+     * @param uint32 Cooldown : The amount of milliseconds the Spell will be on cooldown for.
```
Comment (55Honey) [inline src/LuaEngine/PlayerMethods.h:4207]: It's in seconds

File: src/LuaEngine/PlayerMethods.h [hunk 2]
```diff
@@ -4199,5 +4199,59 @@ namespace LuaPlayer
...
+     * @param bool SendToClient = true : Send cooldown packet to the client.
+     * @param ObjectGuid ItemGuid = 0 : Item Guid if the spell is attached to an item.
+     * @param bool ForceSendToSpectators = false : Should the cooldown packet be force sent to spectators.
+     */
+    int AddSpellCooldown(lua_State* L, Player* player)
+    {
+#ifdef AZEROTHCORE
+        uint32 spellId = Eluna::CHECKVAL<uint32>(L, 2);
+        uint32 cooldown = Eluna::CHECKVAL<uint32>(L, 3) * IN_MILLISECONDS;
```
Comment (55Honey) [inline src/LuaEngine/PlayerMethods.h:4216]: seconds again

File: src/LuaEngine/PlayerMethods.h [hunk 3]
```diff
@@ -4199,5 +4199,59 @@ namespace LuaPlayer
...
+        }
+#endif
+        return 0;
+    }
+
+    /**
+     * Modifys a spell cooldown for the [Player] if a cooldown for the spell exists.
+     *
+     * @param uint32 SpellId : Spell Id to add cooldown to.
+     * @param int32 Cooldown : The amount of milliseconds the Spell cooldown will be changed by.
```
Comment (55Honey) [inline src/LuaEngine/PlayerMethods.h:4244]: Once more, seconds

File: src/LuaEngine/PlayerMethods.h [hunk 4]
```diff
@@ -4199,5 +4199,59 @@ namespace LuaPlayer
...
+     */
+    int ModifySpellCooldown(lua_State* L, Player* player)
+    {
+#ifdef AZEROTHCORE
+        uint32 spellId = Eluna::CHECKVAL<uint32>(L, 2);
+        int32 cooldown = Eluna::CHECKVAL<int32>(L, 3) * IN_MILLISECONDS;
```
Comment (55Honey) [inline src/LuaEngine/PlayerMethods.h:4250]: Same here. Seconds.

**Comment by 55Honey** [comment]
Thanks for the adjustments!

```lua
local SPELL_HEARTHSTONE = 8690
local ITEM_HEARTHSTONE = 6948
local SPELL_HOT = 17928

local function command( event, player, command, chathandler )
    if player then
        if command == 'cddown' then
            player:SendUnitSay('1', 0)
            player:ModifySpellCooldown(SPELL_HEARTHSTONE,-600)
        elseif command == 'cdup' then
            player:ModifySpellCooldown(SPELL_HEARTHSTONE,600)
            player:SendUnitSay('2', 0)
        elseif command == 'addcd' then
            local HearthstoneGuid = player:GetItemByEntry( ITEM_HEARTHSTONE ):GetGUID()
            player:AddSpellCooldown(SPELL_HEARTHSTONE, 1800, true, false, HearthstoneGuid)
            player:SendUnitSay('3', 0)
        end
    end
end

RegisterPlayerEvent(42, command)
```
I have changed the code as shown above. 
`addcd` results in a 30-second timer being shown at the hearthstone. During these 30 seconds, the client correctly behaves when clicking the hearthstone again and states it is on cooldown.
Once the 30 seconds have passed, the behavior is the same as with the initial commit.

Is there an issue with seconds/milliseconds being involved with one function but not the other?

**Comment by Terrorblade** [comment]
From what I can tell in the code they both take milliseconds and what it sounds like it that the server is thinking its still on cd while the client is saying its not.

**Comment by 55Honey** [comment]
For whatever reason, the client always assumes a CD for the hearthstone of 30 seconds. Even when I enter 20 or 50 instead of 1800 as the duration for AddSpellCooldown.

With 50, I can use the hearthstone right after the 30seconds client side CD has expired. So you seem to be right about milliseconds. But I still can't make sense of it.

**Comment by 55Honey** [comment]
@Terrorblade feel free to re-open it whenever you want to look into it. Closing it for now.

**Pull request closed without merging**